### ISSUE TITLE
Partial Sequence Design?

### ISSUE THREAD

**Issue by spark157** [gjoni/trDesign#2]
Hello,

I was wondering if it is possible to do a partial sequence design? What I'm thinking is that I would like to start with a known backbone (from a PDB file) and then mask out some of the residues and allow the model to predict only those. 

I can see there is some functionality to 'spike' the seq_start but I think I'm thinking of something a bit different. Also, it would probably be necessary to backprop and update only the residues covered by the mask.

Like to hear your thoughts or ideas as I start thinking about this.

Thanks!

Scott

**Comment by spark157** [comment]
In looking at bit more at the code I'm wondering if I could pass in my own 'I' matrix that would be 1 hot encoded for the residues I want to keep and then normal(0, 0.01) for those I want the model to design. The loss should be 0 for those residues that are 1hot encoded and therefore match the target sequence and hence no update gets pushed back.  Only the residues that started normal(0,0.01) will have any loss and get updated. 

A follow up though is why on line 264 in models.py are you adding some random noise (I probably no longer want this) and then why on 270/271 are you permuting the the 'I' matrix (indicates only for MSA but seems to be for all) [Update: I realize now this permutation is for the number of sequences so I can see this makes sense.]

Thanks!

Scott

**Comment by sokrypton** [comment]
Technically, if you have a single sequence, even if you "permute" nothing happens 😄 
But, to make it less confusing for future code-readers, I added some logic to disable shuffling unless num > 1.

For partial design, yep, exactly, you can provide your own input to "I". But since this is in log-space, you'll want to multiply the one-hot by a LARGE value so after softmax, they remain one-hot and don't budge during gradient updates (or random addition of noise)

So you can do something like:
`pdb = prep_input("6MRR.pdb", chain="A")`
`_feat = pdb["feat"]`
`_seq = np.eye(20)[AA_to_N(pdb["seq"])]`
`I = np.zeros((1,1,_seq.shape[0],20))`
`I[...,5:20,:] = _seq[5:20] * 1e8`
`design = model.design(inputs={"pdb":_feat[None], "I":I})`

**Comment by spark157** [comment]
Yes - that is working. Thanks.

What does your intuition tell you about how well you think the model will perform on predicting a smaller part of the sequence, say 5%? Would you anticipate it to do well or does it become perhaps more 'random' at the finer level?

Scott
